src/mesh.cpp

```diff
@@ -1852,6 +1989,221 @@ UnstructuredMesh::write(std::string base_filename) const {
 
 #endif
 
+#ifdef LIBMESH
+
+LibMesh::LibMesh(pugi::xml_node node) : UnstructuredMesh(node)
+{
+  initialize();
+}
+
+LibMesh::LibMesh(const std::string& filename)
+{
+  filename_ = filename;
+  initialize();
+}
+
+void LibMesh::initialize()
+{
+  if (!settings::libmesh_init) {
+    fatal_error("Attempting to use an unstructured mesh without a libMesh::LibMeshInit object.");
+  }
+
+  m_ = std::make_unique<libMesh::Mesh>(settings::libmesh_init->comm(), n_dimension_);
+  m_->read(filename_);
+  m_->prepare_for_use();
+
+  // ensure that the loaded mesh is 3 dimensional
+  if(m_->mesh_dimension() != n_dimension_) {
+    fatal_error(fmt::format("Mesh file {} specified for use in an unstructured mesh is not a 3D mesh.", filename_));
+  }
+
+  // create an equation system for storing values
+  eq_system_name_ = fmt::format("mesh_{}_system", id_);
+
+  equation_systems_ = std::make_unique<libMesh::EquationSystems>(*m_);
+  libMesh::ExplicitSystem& eq_sys =
+    equation_systems_->add_system<libMesh::ExplicitSystem>(eq_system_name_);
+
+
+  #ifdef _OPENMP
+  int n_threads = omp_get_max_threads();
+  #else
+  int n_threads = 1;
+  #endif
+
+  for (int i = 0; i < n_threads; i++) {
+      pl_.emplace_back(m_->sub_point_locator());
+      pl_.back()->set_contains_point_tol(FP_COINCIDENT);
+      pl_.back()->enable_out_of_mesh_mode();
+  }
+
+  // store first element in the mesh to use as an offset for bin indices
+  first_element_ = *m_->elements_begin();
+
+  // bounding box for the mesh for quick rejection checks
+  bbox_ = libMesh::MeshTools::create_bounding_box(*m_);
+}
+
+Position
+LibMesh::centroid(int bin) const
+{
+  const auto& elem = this->get_element_from_bin(bin);
+  auto centroid = elem.centroid();
+  return {centroid(0), centroid(1), centroid(2)};
+}
+
+std::string LibMesh::library() const { return "libmesh"; }
+
+int LibMesh::n_bins() const
+{
+  return m_->n_elem();
+}
+
+int LibMesh::n_surface_bins() const
+{
+  int n_bins = 0;
+  for (int i = 0; i < this->n_bins(); i++) {
+    const libMesh::Elem& e = get_element_from_bin(i);
+    n_bins += e.n_faces();
+    // if this is a boundary element, sure to count boundary faces
```

Comment:
Can you elaborate more on this? Only faces on the boundary are counted twice?

Reply:
I'll update the comment to be more clear here. 

All element faces should account for two bins -- a bin for particles that cross the face in each direction. In looping over the mesh elements this happens naturally for interior faces, but faces on the boundary will only be counted once. The inner for loop here increments the number of surface bins for any boundary faces of the element to account for this.

Reply:
Ah, gotcha, that makes sense! The original comment had me thinking that only the boundary faces are counted twice - thanks for clarifying

Reply:
Glad to have revisited this. The `libMesh::Elem::neighbor_ptr_range()` method makes this loop a bit more readable.